mpi-proxy-split/mpi-wrappers/mpi_collective_p2p.c

```diff
@@ -485,39 +485,58 @@ int MPI_Reduce_scatter(const void* sendbuf, void* recvbuf,
 int MPI_Scan(const void* sendbuf, void* recvbuf, int count,
              MPI_Datatype datatype, MPI_Op op, MPI_Comm comm) {
   PROLOG_Comm_rank_size;
+
+  // Get MPI datatype size to allocate buffers
   int root = 0;
   MPI_Aint lower_bound;
   MPI_Aint extent;
   MPI_Type_get_extent(datatype, &lower_bound, &extent);
+  int retval;
+
+  // Deal with buffers in place
   int inplace = (sendbuf == MPI_IN_PLACE);
   if (inplace) {
     sendbuf = recvbuf;
   }
-  if (rank == root) { // root is rank 0
-    int i;
-    char *tmp_sendbuf = (char *)malloc(count * extent);
-    // Gather data into tmp_sendbuf at root
-    MPI_Gather(sendbuf, count, datatype, tmp_sendbuf, count, datatype,
-               root, comm);
-    memcpy(recvbuf, tmp_sendbuf, count * extent);
-    // Initialize tmp_recvbuf from sendbuf at rank 0
-    char *tmp_recvbuf = (char *)malloc(count * extent);
-    memcpy(tmp_recvbuf, tmp_sendbuf, count * extent);
-    for (i = root+1; i < size; i++) {
-      MPI_Reduce_local(tmp_sendbuf + i*extent*count /* inbuf */,
-                       tmp_recvbuf /* inoutbuf */, count, datatype, op);
-      MPI_Send(tmp_recvbuf, count, datatype, i, 0, comm);
+
+  // Root performs reduction
+  if (rank == root) {
+    // Allocate buffer to gather each rank's data, and gather data
+    char *tmp_buf = (char*) malloc(count * extent * size);
+    retval = MPI_Gather(sendbuf, count, datatype, tmp_buf, count, datatype, 0,
+                        comm);
+    if (retval != MPI_SUCCESS) return retval;
+
+    // Matrix of data must be transposed for reduce to be successful
```

Comment:
Can we add more to this comment?  What is the datatype for which we have a matrix that must be transposed?